spec.bs

```diff
@@ -0,0 +1,369 @@
+        1. Continue.
+    1. If |E|.topic-calculation-input-data does not exist:
+        1. Continue.
+    1. <a href=#determine-the-topics-of-a-page-header>Determine</a> the |topics| of the page visit based on |E|.topics-calculation-input-data.
+    1. For each |topic| in |topics|: 
+        1. If |topics-callers|[|topic|] does not exist:
+            1. Initialize |topics-callers|[|topic|] to be an empty list [].
+        1. If |topics-count|[|topic|] does not exist:
+            1. Initialize |topics-count|[|topic|] to 0.
+        1. For each |O| in |E|.topics-caller-origins:
+            1. Append |O| to |topics-callers|[|topic|].
+        1. If E's visit time is after |user topics data start time|:
+            1. Increment |topics-count|[|topic|] by 1.
+  1. Let |top-5-topics| be a list containing the top up to 5 topics in |topics-count|, where topics with more count are retrived first.
+  1. If |top-5-topics| has less than 5 entries:
+      1. Pad |top-5-topics| with random topics from [=taxonomy=].
+  1. Let |top-5-topics-with-caller-origins| be an empty list [].
+  1. For each topic |t| in |top-5-topics|:
+    1. Let |topic-with-caller-origins| be a default struct {topic: |undefined|, caller-origins: []}.
+    1. If |t| is allowed by user preference setting:
+        1. Set |topic-with-caller-origins|.topic = |t|.
+        1. Set |topic-with-caller-origins|.caller-origins = |topics-callers|[|t|].
+    1. Append |topic-with-caller-origins| to |top-5-topics-with-caller-origins|.
+
+  1. Let |epoch| = {id: a 64 bit number distinctive with the ones assigned to past epochs,
+                    taxonomy: [=taxonomy=],
+                    taxonomy-version: [=taxonomy version=],
+                    model-version: [=model version=],
+                    config-version: [=configuration version=],
+                    top-5-topics-with-caller-origins: |top-5-topics-with-caller-origins|, 
+                    time = |now|}.
+
+  1. Append |epoch| to [=user-topics-state=].epochs.
+  1. If [=user-topics-state=].epochs has more than 4 entries, remove the oldest epoch.
+
+  <h2 id="epochs-for-caller-header">Epochs for caller</h2>
+  For a [=topics-caller-context=] |C|, the <dfn>epochs for caller</dfn> method steps are:
+  1. If [=user-topics-state=].epochs is empty:
+      1. Return an empty list [].
+  1. Let |last-epoch-time| = [=user-topics-state=].epochs[-1].time.
+  1. Let |epoch-switch-time-decision-hash| = 64 bit truncation of HMAC-SHA256("epoch-switch-time-decision", [=user-topics-state=].hmac-key, |C|.page.top-level-domain).
+  1. Let |epoch-switch-time-decision-duration| = Seconds(|epoch-switch-time-decision-hash| % 604800) (i.e. 604800 is 7 days in seconds).
+  1. Let |now| be the current time.
+  1. If |now| - |epoch-switch-time-decision-duration| > |last-epoch-time|:
+      1. Let |result| = the last up to 3 epochs from [=user-topics-state=].epochs (i.e. [=user-topics-state=].epochs[-1], [-2], and [-3] for any valid index).
```

Comment:
Variables are block-scoped, so this one will not be visible outside the "If" block: https://infra.spec.whatwg.org/#variables . Declare it outside (initially null, probably) then set it inside these if/else clauses.